k-distribution/src/main/scripts/lib/pyk/kast.py

```diff
@@ -1,216 +1,327 @@
+        return json.dumps(KAst._to_dict(x))
 
+    @final
+    def to_json(self) -> str:
+        return json.dumps(self.to_dict(), sort_keys=True)
 
-def KConstant(label):
-    return KApply(label, [])
+    @staticmethod
+    def from_dict(d: Dict[str, Any]) -> Any:
+        # TODO strengthen return type once the whole AST is represented by classes
+        node = d['node']
+        if node == 'KVariable':
+            return KVariable(name=d['name'])
 
+        if node == 'KToken':
+            return KToken(token=d['token'], sort=d['sort'])
 
-def isKConstant(k):
-    return isKApply(k) and len(k['args']) == 0
+        if node == 'KApply':
+            return KApply(label=d['label'], args=(KAst.from_dict(arg) for arg in d['args']))
 
+        def map_value(x: Any):
+            if type(x) is dict:
+                if 'node' in x:
+                    return KAst.from_dict(x)
+                return x
+            if type(x) is list:
+                return [map_value(e) for e in x]
+            return x
 
-def KSequence(items):
-    return {'node': 'KSequence', 'arity': len(items), 'items': items}
+        return {k: map_value(v) for k, v in d.items()}
 
+    @staticmethod
+    def from_json(s: str) -> 'KAst':
+        return KAst.from_dict(json.loads(s))
 
-def isKSequence(k):
-    return k['node'] == 'KSequence'
 
+class KOuter(KAst, ABC):
+    pass
 
-def KVariable(name):
-    return {'node': 'KVariable', 'name': name, 'originalName': name}
 
+class KInner(KAst, ABC):
+    pass
 
-def isKVariable(k):
-    return k['node'] == 'KVariable'
 
+K: TypeAlias = KInner
 
-def KToken(token, sort):
-    return {'node': 'KToken', 'sort': sort, 'token': token}
 
+@final
+@dataclass(frozen=True)
+class KAtt(KAst):
+    _att: FrozenSet[Tuple[str, Any]]
 
-def isKToken(k):
-    return k['node'] == 'KToken'
+    def __init__(self, att: Mapping[str, Any] = {}):
+        object.__setattr__(self, '_att', frozenset((k, v) for k, v in att.items()))
 
+    @property
+    def att(self) -> Dict[str, Any]:
+        return {k: v for k, v in self._att}
 
-def KRewrite(lhs, rhs):
-    return {'node': 'KRewrite', 'lhs': lhs, 'rhs': rhs}
+    def to_dict(self) -> Dict[str, Any]:
+        return {'node': 'KAtt', 'att': self.att}
 
 
-def isKRewrite(k):
-    return k['node'] == 'KRewrite'
+@final
+@dataclass(frozen=True)
+class KVariable(KInner):
+    name: str
+
+    def __init__(self, name: str):
+        object.__setattr__(self, 'name', name)
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {'node': 'KVariable', 'name': self.name, 'originalName': self.name}
+
+
+@final
+@dataclass(frozen=True)
+class KToken(KInner):
+    token: str
+    sort: str
+
+    def __init__(self, token: str, sort: str):
+        object.__setattr__(self, 'token', token)
+        object.__setattr__(self, 'sort', sort)
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {'node': 'KToken', 'token': self.token, 'sort': self.sort}
```

Comment:
Why don't `KVariable` and `KToken` have `copy` method and `KApply` does?

Reply:
Good point, I'll make sure to add them (and some other utility functions as well).